### PR TITLE
Add backtrack word search in matrix

### PR DESCRIPTION
**Pull request by alexpantyukhin**
### Describe your change:

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [ ] All new algorithms include at least one URL that points to Wikipedia or another similar explanation.
* [ ] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

File: backtracking/word_search.py [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+    len_board_column = len(board[0])
+
+    # Returns the hash key of matrix indexes.
+    def get_point_key(i: int, j: int) -> int:
+        return len_board * len_board_column * i + j
+
+    # Return True if it's possible to search the word suffix
+    # starting from the word_index.
+    def exits_word(i: int,
+                   j: int,
```
[the remark this answers is not in the document]
Comment (alexpantyukhin) [reply backtracking/word_search.py]: fixed

File: backtracking/word_search.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+    len_board = len(board)
+    len_board_column = len(board[0])
+
+    # Returns the hash key of matrix indexes.
+    def get_point_key(i: int, j: int) -> int:
+        return len_board * len_board_column * i + j
+
+    # Return True if it's possible to search the word suffix
+    # starting from the word_index.
+    def exits_word(i: int,
```
[the remark this answers is not in the document]
Comment (alexpantyukhin) [reply backtracking/word_search.py]: fixed

File: backtracking/word_search.py [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+            "The word parameter should be a string of length greater than 0."
+        )
+
+    traverts_directions = [(0, 1), (0, -1), (-1, 0), (1, 0)]
+    len_word = len(word)
+    len_board = len(board)
+    len_board_column = len(board[0])
+
+    # Returns the hash key of matrix indexes.
+    def get_point_key(row: int, column: int) -> int:
```
[the remark this answers is not in the document]
Comment (alexpantyukhin) [reply backtracking/word_search.py:117]: fixed

File: backtracking/word_search.py [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+        return len_board * len_board_column * row + column
+
+    # Return True if it's possible to search the word suffix
+    # starting from the word_index.
+    def exits_word(
```
[the remark this answers is not in the document]
Comment (alexpantyukhin) [reply backtracking/word_search.py:122]: fixed

File: backtracking/word_search.py [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+    False
+    >>> word_exists([["A"]], "A")
+    True
+    >>> word_exists([["A","A","A","A","A","A"],\
+                     ["A","A","A","A","A","A"],\
+                     ["A","A","A","A","A","A"],\
+                     ["A","A","A","A","A","A"],\
+                     ["A","A","A","A","A","B"],\
+                     ["A","A","A","A","B","A"]],\
+                    "AAAAAAAAAAAAABB")
```
Comment (cclauss) [inline backtracking/word_search.py:53]: Line continuation in doctests is done with ...
```suggestion
    >>> word_exists([["A","A","A","A","A","A"],
    ...              ["A","A","A","A","A","A"],
    ...              ["A","A","A","A","A","A"],
    ...              ["A","A","A","A","A","A"],
    ...              ["A","A","A","A","A","B"],
    ...              ["A","A","A","A","B","A"]],
    ...             "AAAAAAAAAAAAABB")
```

File: backtracking/word_search.py [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+
+Word:
+"ABCCED"
+
+Result:
+True
+
+Implementation notes: Use backtracking approach.
+For each point need to check neigh neighbours
+and try to find left suffix of the word.
```
Comment (cclauss) [inline backtracking/word_search.py:30]:
```suggestion
At each point, check all neighbors to try to find the next letter of the word.
```

File: backtracking/word_search.py [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+        if board[row][column] != word[word_index]:
+            return False
+
+        if word_index == len_word - 1:
+            return True
+
+        for direction in traverts_directions:
+            next_i = row + direction[0]
+            next_j = column + direction[1]
+            if not (0 <= next_i < len_board and 0 <= next_j < len_board_column)
```
[the remark this answers is not in the document]
Comment (cclauss) [reply backtracking/word_search.py:134]: Please add a colon (:) at the the end.

[the remark this answers is not in the document]
Comment (cclauss) [reply backtracking/word_search.py:134]:
```suggestion
            if not (0 <= next_i < len_board and 0 <= next_j < len_board_column):
```

**Review by cclauss — APPROVED** [review]

**Pull request merged**